debian/debbindiff.1.rst
============
 debbindiff
============

-----------------------------------------------------------
highlight differences between two builds of Debian packages
-----------------------------------------------------------

:Author: lunar@debian.org
:Copyright: GPL-3+
:Manual section: 1
:Manual group: Debian

SYNOPSIS
========

  debbindiff [-h] [--version] [--debug] [--html output] [--text output] [--max-report-size bytes] [--css url] file1 file2

DESCRIPTION
===========

debbindiff was designed to easily compare two builds of the same Debian
package, and understand their differences.

It can be scripted through error codes, and a report can be produced
with the detected differences. The report can be text or HTML.
When no type of report has been selected, debbindiff defaults
to write a text report on the standard output.

debbindiff was written as part of the “reproducible builds” Debian
project.

OPTIONS
=======

-h, --help               show this help message and exit
--version                show program's version number and exit
--debug                  display debug messages
--html output            write HTML report to given file
                         (use - for standard output)
--text output            write plain text report to given file
                         (use - for standard output)
--max-report-size bytes  maximum bytes written in report
--css url                link to an extra CSS for the HTML report

EXIT STATUS
===========

Exit status is 0 if inputs are the same, 1 if different, 2 if trouble.

SEE ALSO
========

* `<https://wiki.debian.org/ReproducibleBuilds>`


debian/changelog
debbindiff (18) unstable; urgency=low

  [ Reiner Herrmann ]
  * Use zip comparator for pk3 files.
  * Use text comparator for info files.
  * Drop mime type for info files.
  * Remove limit of diff output again.

  [ Jérémy Bobbio ]
  * Assume tar member names are UTF-8 encoded. This allows debbindiff to
    process drmips.
  * Write a text report on stdout as default behavior.
  * Allow both --html and --text to produce output in one run.

 -- Jérémy Bobbio <lunar@debian.org>  Sat, 09 May 2015 11:29:46 +0200

debbindiff (17) unstable; urgency=low

  [ Jérémy Bobbio ]
  * Add support for ISO9660 images.
  * Catch a few more initrds as such.
  * Reimplement are_same_binaries() using cmp.

  [ Reiner Herrmann ]
  * Use gzip comparator for .dz (dictzip) files.
  * Added isoinfo to list of tool providers.

  [ Yasushi SHOJI ]
  * Disable 'Device' section diff in stat output. (Closes: #783792)

 -- Jérémy Bobbio <lunar@debian.org>  Thu, 30 Apr 2015 19:29:26 +0200

debbindiff (16) unstable; urgency=low

  [ Reiner Herrmann ]
  * Remove temporary directory when exception is raised. (Closes: #782253)
  * Fix extraction when Zip member contain "../" in their path.

  [ Jérémy Bobbio ]
  * Fail with a better error message when unable to create a Queue.
    (Closes: #782551)

 -- Jérémy Bobbio <lunar@debian.org>  Thu, 16 Apr 2015 11:07:58 +0200

debbindiff (15) unstable; urgency=low

  * Terminate xxd if its output was too long.
  * Handle broken zip files by falling back on binary comparison.
  * Fix an over-matching jar/war filename regexp.
  * Fix .hi comparator.
  * Fix some file descriptor leaks.

 -- Jérémy Bobbio <lunar@debian.org>  Sat, 04 Apr 2015 18:30:48 +0200

debbindiff (14) unstable; urgency=low

  * Remove diffutils from Recommends as it is Essential:yes.
  * Fallback to Python hexlify if xxd is not available.
  * Decode msgunfmt output using the actual PO charset.
  * Stop buffering output of most external commands and
    send it to diff as it comes. (Closes: #781373)
  * Stop feeding input to diff after a certain amount of lines,
    as GNU diff cannot cope with arbitrary large output.
  * Fix newline and tab handling in HTML presenter.

 -- Jérémy Bobbio <lunar@debian.org>  Thu, 02 Apr 2015 16:16:59 +0200

debbindiff (13) unstable; urgency=low

  [ Helmut Grohne ]
  * Fix several imports for RequiredToolNotFound.

  [ Jérémy Bobbio ]
  * Remove dead code from HTML presenter.
  * Fix tab handling in HTML presenter.
  * Swallow stderr when running showttf. (Closes: #781374)
  * Swallow stderr for readelf and objdump.
  * Use pipes instead of temp files to feed diff.
  * Stop processing diff output after too many lines.
  * Use LINESIZE constant directly in HTML presenter.
  * Better handle non-printable characters in HTML presenter.
  * Cut long lines in HTML presenter.

 -- Jérémy Bobbio <lunar@debian.org>  Sat, 28 Mar 2015 18:53:29 +0100

debbindiff (12) unstable; urgency=low

  [ Helmut Grohne ]
  * Attempt at fixing unicode issues with --text. (Closes: #778641)

  [ Jérémy Bobbio ]
  * Try with utf-8 encoding when comparing text files and no common
    encoding has been detected.
  * Perform content comparison when finding differences instead of
    waiting for the presentation phases. (Closes: #781280)
  * Instead of using vim, the HTML view is created from diff output.
    (Closes: #772029, #779476)

 -- Jérémy Bobbio <lunar@debian.org>  Fri, 27 Mar 2015 18:42:26 +0100

debbindiff (11) unstable; urgency=low

  * Use latin-1 to decode showttf output.
  * Normalize locale environment variables on startup and subsequently use
    UTF-8 to decode output of most external tools. (Closes: #780863)
  * Error out when encoding is not Unicode-compatible for text output.
    (Closes: #778641)

 -- Jérémy Bobbio <lunar@debian.org>  Mon, 23 Mar 2015 22:00:15 +0100

debbindiff (10) unstable; urgency=low

  [ Chris Lamb ]
  * SVGZ files are gzip files.
  * Use gzip comparator for .tgz files.

  [ Reiner Herrmann ]
  * Use zip comparator for Java web archives (.war).
  * Use zip comparator for epub files.
  * Don't fill up diffs with dots.
  * Add support for squashfs images.

  [ Jérémy Bobbio ]
  * Output a nicer name for sub-directories.
  * Also compare stat, acl, and xattrs when comparing directories.
  * Check for availability of tools required by comparators (original work by
    Reiner Herrmann). Most packages in Depends are now in Recommends.
  * Ask objdump to display more sections.
  * Drop stderr output when running cpio -t.
  * Attempt to sort out multiple decoding and encoding issues.
    (Closes: #778641)
  * Fix source path when handling gzip, bzip2 and xz. (Closes: #779474)
  * Don't crash when two .changes don't contain the same files.
    (Closes: #779391)
  * Properly split lines when comparing .changes.
  * Add newlines if needed in text output.
  * Remove most table borders in HTML output. (Closes: #779475)
  * Properly close <span/> in HTML output.
  * Add anchors in HTML output. (Closes: #778537)
  * Update debian/copyright.

 -- Jérémy Bobbio <lunar@debian.org>  Fri, 20 Mar 2015 13:52:44 +0100

debbindiff (9) unstable; urgency=medium

  [ Chris Lamb ]
  * Add rpm-common to Build-Depends so that we can “import rpm" during build.
  * Code improvements improvements.

  [ Jérémy Bobbio ]
  * Add plain text output for differences. Thanks Helmut Grohne for the
    original patch. (Closes: #778423)
  * Exit with 2 in case of errors. (Closes: #774983)
  * Properly output diffs when one of the file is empty.
  * Add support for comparing directories.
  * Update debian/copyright.

 -- Jérémy Bobbio <lunar@debian.org>  Sun, 15 Feb 2015 12:13:46 +0100

debbindiff (8) unstable; urgency=medium

  [ Peter De Wachter ]
  * Use `diff -a` so debbindiff doesn't abort on binary files.

  [ Jérémy Bobbio ]
  * Verify that version in debian/changelog matches debbindiff version.
  * Update debian/copyright.
  * Recognize text/plain files.

  [ Reiner Herrmann ]

debbindiff.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# debbindiff: highlight differences between two builds of Debian packages
#
# Copyright © 2014-2015 Jérémy Bobbio <lunar@debian.org>
#
# debbindiff is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# debbindiff is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with debbindiff.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import argparse
from contextlib import contextmanager
import logging
import codecs
import os
import sys
import traceback
from debbindiff import logger, VERSION
import debbindiff.comparators
from debbindiff.presenters.html import output_html
from debbindiff.presenters.text import output_text


def create_parser():
    parser = argparse.ArgumentParser(
        description='Highlight differences between two builds '
                    'of Debian packages')
    parser.add_argument('--version', action='version',
                        version='debbindiff %s' % VERSION)
    parser.add_argument('--list-tools', nargs=0, action=ListToolsAction,
                        help='show external tools required and exit')
    parser.add_argument('--debug', dest='debug', action='store_true',
                        default=False, help='display debug messages')
    parser.add_argument('--html', metavar='output', dest='html_output',
                        help='write HTML report to given file (use - for stdout)')
    parser.add_argument('--text', metavar='output', dest='text_output',
                        help='write plain text output to given file (use - for stdout)')
    parser.add_argument('--max-report-size', metavar='BYTES',
                        dest='max_report_size', type=int,
                        help='maximum bytes written in report')
    parser.add_argument('--css', metavar='url', dest='css_url',
                        help='link to an extra CSS for the HTML report')
    parser.add_argument('file1', help='first file to compare')
    parser.add_argument('file2', help='second file to compare')
    return parser


@contextmanager
def make_printer(path):
    if path == '-':
        output = sys.stdout
    else:
        output = codecs.open(path, 'w', encoding='utf-8')
    def print_func(*args, **kwargs):
        kwargs['file'] = output
        print(*args, **kwargs)
    yield print_func
    if path != '-':
        output.close()


class ListToolsAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        from debbindiff import tool_required, RequiredToolNotFound
        print("External tools required:")
        print(', '.join(tool_required.all))
        print()
        print("Available in packages:")
        print(', '.join(sorted(set([RequiredToolNotFound.PROVIDERS[k]["debian"] for k in tool_required.all]))))
        sys.exit(0)


def set_locale():
    """Normalize locale so external tool gives us stable and properly
    encoded output"""

    for var in ['LANGUAGE', 'LC_ALL']:
        if var in os.environ:
            del os.environ[var]
    for var in ['LANG', 'LC_NUMERIC', 'LC_TIME', 'LC_COLLATE', 'LC_MONETARY',
                'LC_MESSAGES', 'LC_PAPER', 'LC_NAME', 'LC_ADDRESS',
                'LC_TELEPHONE', 'LC_MEASUREMENT', 'LC_IDENTIFICATION']:
        os.environ[var] = 'C'
    os.environ['LC_CTYPE'] = 'C.UTF-8'


def main():
    parser = create_parser()
    parsed_args = parser.parse_args(sys.argv[1:])
    if parsed_args.debug:
        logger.setLevel(logging.DEBUG)
    set_locale()
    differences = debbindiff.comparators.compare_files(
        parsed_args.file1, parsed_args.file2)
    if len(differences) > 0:
        if parsed_args.html_output:
            with make_printer(parsed_args.html_output) as print_func:
                output_html(differences, css_url=parsed_args.css_url, print_func=print_func,
                            max_page_size=parsed_args.max_report_size)
        if (parsed_args.text_output and parsed_args.text_output != parsed_args.html_output) or not parsed_args.html_output:
            with make_printer(parsed_args.text_output or '-') as print_func:
                output_text(differences, print_func=print_func)
        return 1
    return 0

if __name__ == '__main__':
    try:
        sys.exit(main())
    except (Exception, KeyboardInterrupt):
        traceback.print_exc()
        sys.exit(2)
